File: JANUARY/08-01-2025/1458.ts
function maxDotProduct(nums1: number[], nums2: number[]): number {
    const n = nums1.length, m = nums2.length;
    const NEG = -1e9;
    const dp: number[][] = Array.from({ length: n + 1 }, () => Array(m + 1).fill(NEG));

    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= m; j++) {
            const prod = nums1[i - 1] * nums2[j - 1];
            dp[i][j] = Math.max(
                prod,
                dp[i - 1][j],
                dp[i][j - 1],
                dp[i - 1][j - 1] + prod
            );
        }
    }
    return dp[n][m];
}

File: JANUARY/08-01-2025/1458.rb
# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Integer}
def max_dot_product(nums1, nums2)
  n, m = nums1.length, nums2.length
  neg = -1_000_000_000
  dp = Array.new(n + 1) { Array.new(m + 1, neg) }

  (1..n).each do |i|
    (1..m).each do |j|
      prod = nums1[i - 1] * nums2[j - 1]
      dp[i][j] = [
        prod,
        dp[i - 1][j],
        dp[i][j - 1],
        dp[i - 1][j - 1] + prod
      ].max
    end
  end

  dp[n][m]
end